### PR TITLE
[elixir] feat: make archive/error jobs more error proof

### PR DESCRIPTION
**Pull request by grejdi-mbta**
This PR modularizes the copy and delete functionality of the archive and error workers. It also adds checks for various states  prior to copying, to make adjustments to logic.

### CODE REVIEW

File: ex_cubic_ingestion/test/support/ex_aws.ex [hunk 1]
```diff
@@ -6,107 +6,43 @@ defmodule MockExAws do
...
+      end) ++
+        [
+          "cubic/dmap/sample/already_copied_but_source_not_deleted.csv.gz",
+          "cubic/dmap/sample/copy_and_delete.csv.gz",
+          "cubic/dmap/sample/20220101.csv"
+        ]
 
     # deleting object
-    if Enum.member?(valid_paths, op.path) do
-      {:ok,
```
Comment (grejdi-mbta) [inline ex_cubic_ingestion/test/support/ex_aws.ex:17]: These information was not being used in tests, so I simplified them.

File: ex_cubic_ingestion/lib/ex_aws/helpers.ex [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+  @spec copy_and_delete(module(), String.t(), String.t(), String.t(), String.t()) :: :ok | :error
+  def copy_and_delete(lib_ex_aws, source_bucket, source_key, destination_bucket, destination_key) do
+    {source_status, _req_response} =
+      lib_ex_aws.request(ExAws.S3.head_object(source_bucket, source_key))
+
+    {destination_status, _req_response} =
+      lib_ex_aws.request(ExAws.S3.head_object(destination_bucket, destination_key))
+
+    cond do
+      # if source and destination exist, just delete the source
```
Comment (paulswartz) [inline ex_cubic_ingestion/lib/ex_aws/helpers.ex:21]: question: should this also do any work to ensure that the source and destination are the same file?

File: ex_cubic_ingestion/lib/ex_cubic_ingestion/workers/archive.ex [hunk 3]
```diff
@@ -37,15 +37,22 @@ defmodule ExCubicIngestion.Workers.Archive do
...
+             archive_bucket,
+             destination_key
+           ) do
+        :ok ->
+          "archived"
 
-    # delete load file from incoming bucket
-    lib_ex_aws.request!(ExAws.S3.delete_object(incoming_bucket, source_key))
+        _error ->
+          "archived_unknown"
```
Comment (paulswartz) [inline ex_cubic_ingestion/lib/ex_cubic_ingestion/workers/archive.ex:52]: suggestion: why not treat the worker as having failed, and have Oban retry the job?

File: ex_cubic_ingestion/lib/ex_aws/helpers.ex (hunk 2 above)
Comment (grejdi-mbta) [reply ex_cubic_ingestion/lib/ex_aws/helpers.ex:21]: Feels like overkill.

File: ex_cubic_ingestion/lib/ex_aws/helpers.ex [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+      source_status == :error and destination_status == :ok ->
+        :ok
+
+      # no source object, so error
+      source_status == :error ->
+        :error
+
+      # do the copy and delete
+      true ->
+        lib_ex_aws.request!(
```
[the remark this answers is not in the document]
Comment (grejdi-mbta) [reply ex_cubic_ingestion/lib/ex_aws/helpers.ex:37]: makes sense. let me make some adjustments.

File: ex_cubic_ingestion/lib/ex_cubic_ingestion/workers/archive.ex (hunk 3 above)
[the remark this answers is not in the document]
Comment (grejdi-mbta) [reply ex_cubic_ingestion/lib/ex_cubic_ingestion/workers/archive.ex:52]: No. Only 'archiving' or 'erroring' states which suggest the job failed right before updating the status. https://github.com/mbta/data_platform/pull/61/files#diff-27132d9d476fc515d74183d3cb1d5ec48316c4a01b2f7521128ddfab9a788342R23

Comment (grejdi-mbta) [reply ex_cubic_ingestion/lib/ex_cubic_ingestion/workers/archive.ex:52]: If something goes wrong such as the permissions issue, retrying with Oban would result in the same outcome. I guess if schedule the retry for the next day, then maybe we would have fixed whatever happened, but still doesn't seem like enough time.

Comment (paulswartz) [reply ex_cubic_ingestion/lib/ex_cubic_ingestion/workers/archive.ex:52]: That's fair. I still think that we should mark the Oban job as failed, even if we don't retry it automatically.

File: ex_cubic_ingestion/lib/ex_aws/helpers.ex (hunk 2 above)
Comment (grejdi-mbta) [reply ex_cubic_ingestion/lib/ex_aws/helpers.ex]: Added an Asana ticket: https://app.asana.com/0/1201701089427502/1202495994757255

File: ex_cubic_ingestion/lib/ex_aws/helpers.ex (hunk 4 above)
[the remark this answers is not in the document]
Comment (grejdi-mbta) [reply ex_cubic_ingestion/lib/ex_aws/helpers.ex]: Updated. Let me know what you think.

File: ex_cubic_ingestion/lib/ex_aws/helpers.ex [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+  Central place for useful functions that do ExAws work.
+  """
+
+  @doc """
+  Performs checks, before copying the source object to a destination object
+  and deleting the source object. First is the check for if they copy has
+  already occurred. And the second check is for existance source object to
+  prevent running a copy/delete on a non-existent source.
+  """
+  @spec copy_and_delete(module(), String.t(), String.t(), String.t(), String.t()) :: :ok | :error
```
[the remark this answers is not in the document]
Comment (grejdi-mbta) [reply ex_cubic_ingestion/lib/ex_aws/helpers.ex]: Done.

File: ex_cubic_ingestion/lib/ex_cubic_ingestion/workers/archive.ex [hunk 6]
```diff
@@ -37,17 +37,22 @@ defmodule ExCubicIngestion.Workers.Archive do
...
+    if move_status == :ok do
+      CubicLoad.update(load_rec, %{status: "archived"})
 
-    CubicLoad.update(load_rec, %{status: "archived"})
-
-    :ok
+      :ok
+    else
+      {:error, move_req_response}
+    end
```
[the remark this answers is not in the document]
Comment (grejdi-mbta) [reply ex_cubic_ingestion/lib/ex_cubic_ingestion/workers/archive.ex:54]: Updated. Always looking for more Elixir-y ways to write Elixir code.

**Review by paulswartz — APPROVED** [review]
🍰

**Pull request merged**